### PR TITLE
[flexsn] Add inductor training backend tuning

### PR DESCRIPTION
**Pull request by fangwei123456**
## Summary
This PR lands the FlexSN inductor training backend work and the follow-up tuning we validated on the CUDA server.

## What changed
- add the FlexSN inductor custom-op training path and HOP/scan integration files
- add training final-state forward/backward specializations and state-template handling fixes
- bucket backward autotune by NCL to better match real training workloads
- add offline benchmarking scripts for VGG training workloads
- add regression tests for training final-state behavior, state-template handling, and backward bucket boundaries

## Why
The goal was to push FlexSN training performance as far as the current generic backend could reasonably go, while keeping the path correct and testable. The last round of work focused on real training workloads rather than inference-only tuning.

## Validation
Validated on the CUDA server with targeted tests:
- `PYTHONPATH=. pytest test/activation_based/test_flex_sn_inductor.py -k 'mismatched_example_state_shape or zero_state_materialization_uses_registered_state_specs' -q`
- `PYTHONPATH=. pytest test/activation_based/test_flex_sn_inductor.py -k 'training_final_state_matches_full_training or training_final_state_kernel_only_builds_when_it_reduces_saved_sequences' -q`
- `PYTHONPATH=. pytest test/activation_based/test_flex_sn_inductor.py -k 'backward_ncl_bucket_boundaries or training_final_state_matches_full_training or backward_final_state_kernel_handles_small_t_large_token_workloads' -q`

Benchmarked on an RTX 4090 with `CUDA_VISIBLE_DEVICES=1`:
- ImageNet `spiking_vgg11_bn`: default `59.43 ms` vs fallback `60.70 ms`
- ImageNet `spiking_vgg16_bn`: default `107.01 ms` vs fallback `110.18 ms`
- CIFAR `spiking_vgg11_bn`: default `23.92 ms` vs fallback `24.88 ms`

## Notes
At this point the easy backend-level wins appear mostly exhausted; further large gains will likely require less-generic specialization rather than more parameter tuning.

## Summary by CodeRabbit

* **New Features**
  * Added an "inductor" execution backend with final-state execution, optional zero-state elision, segmented execution, and NCL-bucket-aware backward handling for improved GPU performance.
  * New GPU benchmarking tools that produce per-workload and per-bucket latency/throughput summaries.

* **Tests**
  * Extensive CUDA and compile-mode tests validating backend equivalence, gradients, final-state semantics, tracing/AOTAutograd integration, and performance.

### CODE REVIEW

**Comment by fangwei123456** [comment]
@gemini @coderabbit @codex please review this PR. I’d especially like feedback on the FlexSN inductor training path, the state-template handling fixes, and the backward autotune bucketing for real training workloads.

**Comment by fangwei123456** [comment]
Review fixes are pushed in `062a0c04`.

Looping in reviewers again for another pass:
- @gemini
- @coderabbit
- @codex

I also re-ran the targeted CUDA regression tests on the updated branch.

**Comment by fangwei123456** [comment]
Follow-up review fixes are now pushed in `8680f2c1`.

Addressed this round:
- fallback inference now preserves final-state semantics when the specialized final-state kernel is unavailable
- backward final-state threshold gating no longer enables specialization when only one knob is disabled
- HOP eager scan normalizes single-tensor returns before arity checks
- final-state Triton builders now handle `num_states == 0` without generating invalid signatures
- `benchmark_vgg_inductor.py` now exercises training instead of inference-only timing
- backward runtime template args no longer allocate full-sized tensors

Re-ran targeted CUDA tests on the updated branch:
- `8 passed` in `test_flex_sn_inductor.py` for the relevant follow-up cases

Looping in reviewers again:
- @gemini
- @coderabbit
- @codex

**Comment by fangwei123456** [comment]
Latest review fixes are pushed in `64c8dea6`.

Addressed this round:
- `hop.py`: allow trailing lifted (non-scanned) tensor operands and thread them through each timestep call
- `template.py`: build backward-final-state input signatures from parts so `num_outputs == 0` does not produce a leading comma
- tests: strengthened zero-state materialization coverage with a mixed-device CUDA case, and updated HOP arity coverage for lifted operands

Re-ran targeted CUDA regressions on the updated branch:
- `11 passed` in `test_flex_sn_inductor.py` for the relevant HOP/inductor follow-up cases

Looping in reviewers again:
- @gemini-code-assist
- @coderabbit
- @codex

**Comment by fangwei123456** [comment]
Addressed the latest actionable review feedback in `ab82c1ae`.

What changed:
- added a graph-derived fingerprint to FlexSN inductor kernel names so distinct `lambda` / `functools.partial` wrappers do not collide in Triton cache naming
- added a regression test covering same-name / different-graph kernel naming behavior

Validation:
- remote CUDA regression on the latest branch head passed:
  - `PYTHONPATH=. /home/qwe/user/fangwei/env/pytorch/bin/pytest test/activation_based/test_flex_sn_inductor.py -k 'kernel_names_include_graph_fingerprint or hop_matches_manual_loop_with_lifted_tensor or hop_accepts_single_tensor_return or zero_state_materialization_overrides_registered_device_at_runtime or compile_fullgraph_lowerable_while_loop_matches_eager or training_final_state_matches_full_training or inference_without_final_state_kernel_keeps_final_state_semantics or stateless_final_state_kernel_builds' -q`
  - `7 passed, 1 skipped`

@gemini-code-assist @coderabbit @codex

**Comment by fangwei123456** [comment]
Addressed the latest actionable review feedback in `ad3398b8`.

What changed:
- `eager_scan` and `eager_scan_final_state` now reject empty input sequences with a clear `ValueError` instead of falling through to `torch.stack([])`
- added a regression test covering the `T == 0` contract for both scan variants

Validation on the remote CUDA environment:
- `PYTHONPATH=. /home/qwe/user/fangwei/env/pytorch/bin/pytest test/activation_based/test_flex_sn_inductor.py -k 'hop_rejects_empty_input_sequence or hop_matches_manual_loop_with_lifted_tensor or hop_accepts_single_tensor_return or kernel_names_include_graph_fingerprint or compile_fullgraph_lowerable_while_loop_matches_eager' -q`
- `4 passed, 1 skipped`

The remaining older unresolved threads I see on the PR are stale UI state: the fallback inference/state-sequence handling and backward template-allocation concerns are already covered by the current branch code.

**Comment by fangwei123456** [comment]
Latest fix is in `ad3398b8` and remote regression passed.

@gemini-code-assist @coderabbit @codex

**Comment by fangwei123456** [comment]
Addressed the latest CodeRabbit lint finding in `70289d6f` by replacing the test-local lambda assignments with named `def`s in `test_kernel_names_include_graph_fingerprint`.

Validation:
- remote regression still passes for the affected HOP/kernel-name tests:
  - `3 passed`

@gemini-code-assist @coderabbit @codex

**Comment by fangwei123456** [comment]
Addressed the latest regression-test review feedback in `2c52b677`.

What changed:
- strengthened `test_inductor_inference_without_final_state_kernel_keeps_final_state_semantics` to compare fallback outputs and final states against the normal path, not just shape
- reset the spiking model between eager and compiled runs in `test_compile_fullgraph_lowerable_while_loop_matches_eager` so both paths start from the same initial neuron state

Validation on the remote CUDA environment:
- `PYTHONPATH=. /home/qwe/user/fangwei/env/pytorch/bin/pytest test/activation_based/test_flex_sn_inductor.py -k 'inference_without_final_state_kernel_keeps_final_state_semantics or compile_fullgraph_lowerable_while_loop_matches_eager' -q`
- `1 passed, 1 skipped`

@gemini-code-assist @coderabbit @codex
